<?xml version="1.0"?>
<!-- vim: set fenc=utf-8 et sts=4 ft=xml : -->
<launch>
    <arg name="run_cameras"/>
    <arg name="machine"/>
    <arg name="namespace"/>
    <arg name="manager_name"/>
    <arg name="left_guid"/>
    <arg name="right_guid"/>
    <arg name="strobe"/>
    <arg name="run_processing" default="true"/>
    <group if="$(arg run_cameras)">
        <node name="left"
              machine="$(arg machine)"
              pkg="nodelet"
              type="nodelet"
              args="load camera1394/driver $(arg manager_name)"
              ns="/cameras/navigation/$(arg namespace)"
              respawn_delay="5.0"
              respawn="true">
            <param name="guid" type="string" value="$(arg left_guid)"/>

            <param name="reset_on_open" type="bool" value="true"/>
            <param name="frame_id" value="navigation_$(arg namespace)_left_camera"/>
            <param name="video_mode" type="string" value="640x480_mono8"/>
            <param name="bayer_pattern" value="rggb"/>
            <param name="auto_exposure" type="int" value="3"/>
            <param name="exposure" type="int" value="52"/>
            <param name="auto_shutter" type="int" value="2"/>
            <param name="auto_gain" type="int" value="2"/>
            <param name="auto_brightness" type="int" value="2"/>
            <param name="frame_rate" type="double" value="30"/>
            <param name="camera_info_url"
                   type="string"
                   value="file://$(find samplereturn)/calibration/ffmv-$(arg namespace)-left-752.yaml" />
            <remap from="camera" to="left" />
            <remap from="left/camera_params" to="right/camera_params" />
        </node>
        <node name="right"
              machine="$(arg machine)"
              pkg="nodelet"
              type="nodelet"
              args="load camera1394/driver $(arg manager_name)"
              ns="/cameras/navigation/$(arg namespace)"
              respawn_delay="5.0"
              respawn="true">

            <param name="guid" type="string" value="$(arg right_guid)"/>

            <param name="reset_on_open" type="bool" value="true"/>
            <param name="frame_id" value="navigation_$(arg namespace)_left_camera"/>
            <param name="video_mode" type="string" value="640x480_mono8"/>
            <param name="bayer_pattern" value="rggb"/>
            <param name="auto_exposure" type="int" value="3"/>
            <param name="exposure" type="int" value="52"/>
            <param name="auto_shutter" type="int" value="2"/>
            <param name="auto_gain" type="int" value="2"/>
            <param name="auto_brightness" type="int" value="2"/>
            <param name="frame_rate" type="double" value="30"/>
            <param name="camera_info_url"
                   type="string"
                   value="file://$(find samplereturn)/calibration/ffmv-$(arg namespace)-right-752.yaml" />
            <remap from="camera" to="right" />
        </node>

        <!-- emit synchronization error -->
        <node name="image_desync"
              machine="$(arg machine)"
              pkg="samplereturn"
              type="image_desync.py"
              ns="/cameras/navigation/$(arg namespace)">
            <param name="manager_node_name" type="string" value="/cameras/navigation/$(arg namespace)/$(arg manager_name)"/>
            <param name="max_desync" type="double" value="0.033"/>
        </node>
    </group>

    <group if="$(arg run_processing)">
        <!-- Debayer -->
        <node name="debayer"
            machine="$(arg machine)"
            pkg="nodelet"
            type="nodelet"
            args="load image_proc/debayer /cameras/navigation/$(arg namespace)/$(arg manager_name)"
            ns="/cameras/navigation/$(arg namespace)/left"
            respawn_delay="5.0"
            respawn="true">
        </node>
        <node name="debayer"
            machine="$(arg machine)"
            pkg="nodelet"
            type="nodelet"
            args="load image_proc/debayer /cameras/navigation/$(arg namespace)/$(arg manager_name)"
            ns="/cameras/navigation/$(arg namespace)/right"
            respawn_delay="5.0"
            respawn="true">
        </node>

        <!-- Rectify images -->
        <node name="rect"
            machine="$(arg machine)"
            pkg="nodelet"
            type="nodelet"
            args="load image_proc/rectify /cameras/navigation/$(arg namespace)/$(arg manager_name)"
            ns="/cameras/navigation/$(arg namespace)/left"
            output="log"
            respawn_delay="5.0"
            respawn="true">
            <param name="queue_size" type="int" value="1"/>
        </node>
        <node name="rect_color"
            machine="$(arg machine)"
            pkg="nodelet"
            type="nodelet"
            args="load image_proc/rectify /cameras/navigation/$(arg namespace)/$(arg manager_name)"
            ns="/cameras/navigation/$(arg namespace)/left"
            output="log"
            respawn_delay="5.0"
            respawn="true">
            <remap from="image_mono" to="image_color"/>
            <remap from="image_rect" to="image_rect_color"/>
            <param name="queue_size" type="int" value="1"/>
        </node>
        <node name="rect"
            machine="$(arg machine)"
            pkg="nodelet"
            type="nodelet"
            args="load image_proc/rectify /cameras/navigation/$(arg namespace)/$(arg manager_name)"
            ns="/cameras/navigation/$(arg namespace)/right"
            output="log"
            respawn_delay="5.0"
            respawn="true">
            <param name="queue_size" type="int" value="1"/>
        </node>
        <!--
    <node name="rect_color"
          machine="$(arg machine)"
          pkg="nodelet"
          type="nodelet"
          args="load image_proc/rectify /cameras/navigation/$(arg namespace)/$(arg manager_name)"
          ns="/cameras/navigation/$(arg namespace)/right"
          output="log"
          respawn_delay="30.0"
          respawn="true">
        <remap from="image_mono" to="image_color"/>
        <remap from="image_rect" to="image_rect_color"/>
          <param name="queue_size" type="int" value="1"/>
    </node>
    -->

    <!-- Compute Stereo -->
    <node name="stereo_image_disparity"
        machine="$(arg machine)"
        pkg="nodelet"
        type="nodelet"
        args="load stereo_image_proc/disparity $(arg manager_name)"
        ns="/cameras/navigation/$(arg namespace)"
        output="log"
        respawn_delay="5.0"
        respawn="true">
        <param name="correlation_window_size" type="int" value="17"/>
        <param name="min_disparity" type="int" value="0"/>
        <param name="disparity_range" type="int" value="128"/>
        <param name="uniqueness_ratio" type="int" value="5"/>
        <param name="texture_threshold" type="int" value="1000"/>
        <param name="approximate_sync" type="bool" value="true"/>
    </node>
    </group>
</launch>
